### ISSUE TITLE
`RecursionError: maximum recursion depth exceeded` in `utils.resolve_all`

### ISSUE THREAD

**Issue by jtschoonhoven** [jsvine/pdfplumber#638]
## Describe the bug

The function `utils.resolve_all` is recursive, and it can exceed Python's maximum recursion depth when parsing complex Powerpoint files.

## Code to reproduce the problem

```python
with pdfplumber.open(file_obj) as pdf:
    for page_index, page in enumerate(pdf.pages):
        thumbnail = pdf_thumbnails[page_index]
        for link in page.hyperlinks:
            pass
```

## PDF file

I don't have permission to share the PDF file, but the issue should be self-evident.

## Expected behavior

It should not raise a RecursionError.

## Actual behavior

It raises a RecursionError.

## Screenshots

```
  File "/usr/local/lib/python3.8/site-packages/pdfplumber/utils.py", line 141, in resolve_all
    return dict((k, v if k in exceptions else resolve_all(v)) for k, v in x.items())
  File "/usr/local/lib/python3.8/site-packages/pdfplumber/utils.py", line 141, in <genexpr>
    return dict((k, v if k in exceptions else resolve_all(v)) for k, v in x.items())
  File "/usr/local/lib/python3.8/site-packages/pdfplumber/utils.py", line 127, in resolve_all
    resolved = x.resolve()
  File "/usr/local/lib/python3.8/site-packages/pdfminer/pdftypes.py", line 96, in resolve
    return self.doc.getobj(self.objid)
  File "/usr/local/lib/python3.8/site-packages/pdfminer/pdfdocument.py", line 832, in getobj
    log.debug('getobj: objid=%r', objid)
  File "/usr/local/lib/python3.8/logging/__init__.py", line 1433, in debug
    if self.isEnabledFor(DEBUG):
RecursionError: maximum recursion depth exceeded
```

## Environment

- pdfplumber version: 0.6.0
- Python version: 3.8.12
- OS: Ubuntu Linux

## Additional context

It doesn't look easy to rewrite this function to be iterative instead of recursive, so I'm hoping someone cleverer than me has an idea.

**Comment by jsvine** [comment]
Thanks for flagging, @jtschoonhoven. There are at least a couple of possibilities here:

- The PDF's internal references may, indeed, be infinitely recursive
- There may be a bug in parsing the internal references, either in `pdfplumber` or in `pdfminer.six`

Without a specific PDF, it's difficult to diagnose the issue. Are you able to provide a redacted version, using https://github.com/JoshData/pdf-redactor? Alternatively, a non-sensitive simple example that reproduces the issue?

In the meantime, you may also have luck [repairing the PDF using GhostScript](https://superuser.com/questions/278562/how-can-i-fix-repair-a-corrupted-pdf-file).

**Comment by jsvine** [comment]
Closing this issue due to inactivity and lack of issue-reproducing PDF. If someone is able to provide a PDF that raises the same error, I'll reopen this. Thanks.

**Comment by dvgodoy** [comment]
@jsvine, @jtschoonhoven 

I had the same issue described here, and I figured that the problem was that the document was digitally signed using DocuSign. Apparently, DocuSign adds an annotation to the PDF, and this particular type of annotation refers to the document itself, thus triggering an infinite recursion.

Unfortunately I cannot provide the PDF either, but I can tell you the hack I used to solve it.
Since the signature had a particular type "Sig", and there was already a provision to avoid infinite recursion in the `resolve_all` function in `pdfplumber/utils.py`, I modified it slightly to include "Sig" as well:

```python
def resolve_all(x: Any) -> Any:
    """
    Recursively resolves the given object and all the internals.
    """
    if isinstance(x, PDFObjRef):
        resolved = x.resolve()

        # Avoid infinite recursion - inclusing signatures here
        if get_dict_type(resolved) in ["Page", "Sig"]:
            return x

        return resolve_all(resolved)
    elif isinstance(x, (list, tuple)):
        return type(x)(resolve_all(v) for v in x)
    elif isinstance(x, dict):
        exceptions = ["Parent"] if get_dict_type(x) == "Annot" else []
        return {k: v if k in exceptions else resolve_all(v) for k, v in x.items()}
    else:
        return x
```

That solved my issue. Perhaps it can help with yours too.

Best,
Daniel

**Comment by jsvine** [comment]
Hi @dvgodoy, and thank you for this helpful note. The fact that a DocuSign PDF was causing you trouble sounded like a potential clue, so I generated and signed a simple example / test PDF: [docusign-example.pdf](https://github.com/jsvine/pdfplumber/files/9355656/docusign-example.pdf)

Unfortunately (or fortunately!), `pdfplumber` handles the PDF and its annotations process just fine, no errors of any sort. So I wonder if it's something with the way your document was pre- or post-processed. Do you have any additional information you can share?

**Comment by dvgodoy** [comment]
OK, so it is the document that is broken then.

I had to do some digging, but I was able to find the cause of the recursion. I will try to outline it here using the dictionaries I found along the way.

This is the dict corresponding to the annotation/signature:

{'AP': {'N': <PDFObjRef:411>}, 'DR': {}, 'F': 132, 'FT': /'Sig', 'Lock': <PDFObjRef:5>, 'P': <PDFObjRef:4>, 'Rect': [0, 0, 0, 0], 'Subtype': /'Widget', 'T': b'ENVELOPEID_...', 'Type': /'Annot', 'V': <PDFObjRef:412>}

The problem lies in the 'V' key, the object number 412. By the way, I ran your signed document for comparison too. Mine includes a 'Data' key (in bold, below) inside 'Reference', and yours does not.

{'ByteRange': [0, 1666425, 1731961, 4380], 'ContactInfo': b'', 'Contents': b'...', 'Filter': /'Adobe.PPKMS', 'Location': b'', 'M': b"D:...'00'", 'Prop_Build': {'App': {'Name': /b'DocuSign\xae'}}, 'Reason': b'Digitally verifiable PDF exported from[ www.docusign.com](http://www.docusign.com/)', 'Reference': [{**'Data': <PDFObjRef:1>**, 'DigestLocation': [0, 0], 'DigestMethod': /'MD5', 'DigestValue': b'aa', 'TransformMethod': /'FieldMDP', 'TransformParams': {'Action': /'All', 'P': 1, 'Type': /'TransformParams', 'V': /'1.2'}, 'Type': /'SigRef'}], 'SubFilter': /'adbe.pkcs7.detached', 'Type': /'Sig'}

This 'Data' key points to object number 1 below. Notice the 'Fields' key there (in bold), which points to object number 6.

{'AcroForm': {'DA': b'/Helv 0 Tf 0 g ', 'DR': {'Font': {'Helv': <PDFObjRef:7>, 'ZaDb': <PDFObjRef:8>}}, **'Fields': [<PDFObjRef:6>]**, 'SigFlags': 3}, 'DSS': <PDFObjRef:9>, 'Extensions': {'ESIC': {'BaseVersion': /'1.7', 'ExtensionLevel': 5}}, 'Metadata': <PDFObjRef:10>, 'Outlines': <PDFObjRef:404>, 'Pages': <PDFObjRef:12>, 'Type': /'Catalog', 'Version': /'1.5'}

The problem is, object number 6 is the initial annotation/signature. So it goes around and around, triggering the error.

{'AP': {'N': <PDFObjRef:411>}, 'DR': {}, 'F': 132, 'FT': /'Sig', 'Lock': <PDFObjRef:5>, 'P': <PDFObjRef:4>, 'Rect': [0, 0, 0, 0], 'Subtype': /'Widget', 'T': b'ENVELOPEID_...', 'Type': /'Annot', 'V': <PDFObjRef:412>}

In summary:

Annotation/Signature (id: 6) -> Signature itself (id: 412) -> "Data" in signature (id: 1) -> "Fields" = annotation/signature (id: 6)

Unfortunately I have no idea how this signature was created, neither why it has this 'Data' field that is to blame for this issue.
I guess Adobe Acrobat sees that there's something wrong, because it gives me an error message when I open it, but it still opens the document and I can read it.

In my case, I am interested in "true" annotations only, things that the user may have added to the file. I did not even know that signatures/DocuSign were added as annotations, so I simply got rid of them.

Hope this helps. Please let me know if you need any more digging on my part :-)

Best,
Daniel

**Comment by jsvine** [comment]
Thanks @dvgodoy for these details. Quite helpful. There's probably a way to handle the exception more elegantly than `pdfplumber` does now (so that your experience is more like how it is in Acrobat), but I'm going to wait until I have a concrete PDF I can test it against.

**Comment by FallakAsad** [comment]
[dvgodoy](https://github.com/dvgodoy)

do you mind if you tell how did you manage to get dict type? if its Page or Sig?
